lib/_http_parser.js

```diff
@@ -0,0 +1,907 @@
+    this.onBody(data, 0, len);
+  }
+  return len;
+};
+HTTPParser.prototype._executeBodyEOF = function(data) {
+  var len = data.length;
+
+  if (len === 0)
+    return 0;
+
+  this.onBody(data, 0, len);
+  return len;
+};
+HTTPParser.prototype._executeBodyIgnore = function(data) {
+  return 0;
+};
+HTTPParser.prototype._executeError = function(data) {
+  return this._err;
+};
+HTTPParser.prototype.execute = HTTPParser.prototype._executeStartLine;
+HTTPParser.prototype._shouldKeepAlive = function(httpMajor, httpMinor, flags,
+                                                 type, status) {
+  if (httpMajor > 0 && httpMinor > 0) {
+    if ((flags & FLAG_CONNECTION_CLOSE) > 0)
+      return false;
+  } else {
+    if ((flags & FLAG_CONNECTION_KEEP_ALIVE) === 0)
+      return false;
+  }
+  return !this._needsEOF(flags, type, status);
+};
+HTTPParser.prototype._needsEOF = function(flags, type, status) {
+  if (type === REQUEST)
+    return false;
+
+  // See RFC 2616 section 4.4
+  if (status === 204 ||                    // No Content
+      status === 304 ||                    // Not Modified
+      (status >= 100 && status < 200) ||   // 1xx e.g. Continue
+      (flags & FLAG_SKIPBODY) > 0) {       // response to a HEAD request
+    return false;
+  }
+
+  if ((flags & FLAG_CHUNKED) > 0 || this._contentLen !== null)
+    return false;
+
+  return true;
+};
+
+
+
+
+module.exports = HTTPParser;
+
+function indexOfLF(buf, buflen, offset) {
+  while (offset < buflen) {
+    if (buf[offset] === LF)
+      return offset;
+    ++offset;
+  }
+  return -1;
+}
+
+function hexValue(ch) {
+  if (ch > 47 && ch < 58)
+    return ch - 48;
+  ch |= 0x20;
+  if (ch > 96 && ch < 103)
+    return 10 + (ch - 97);
+}
+
+function ltrim(value) {
+  var length = value.length, start;
+  for (start = 0;
+       start < length &&
+       (value.charCodeAt(start) === 32 || value.charCodeAt(start) === 9 ||
+        value.charCodeAt(start) === CR);
+       ++start);
+  return (start > 0 ? value.slice(start) : value);
+}
+
+function rtrim(value) {
+  var length = value.length, end;
+  for (end = length;
+       end > 0 &&
+       (value.charCodeAt(end - 1) === 32 || value.charCodeAt(end - 1) === 9 ||
+        value.charCodeAt(end - 1) === CR);
+       --end);
+  return (end < length ? value.slice(0, end) : value);
+}
+
+function trim(value) {
+  var length = value.length, start, end;
+  for (start = 0;
+       start < length &&
+       (value.charCodeAt(start) === 32 || value.charCodeAt(start) === 9 ||
+        value.charCodeAt(start) === CR);
+       ++start);
+  for (end = length;
+       end > start &&
+       (value.charCodeAt(end - 1) === 32 || value.charCodeAt(end - 1) === 9 ||
+        value.charCodeAt(end - 1) === CR);
+       --end);
+  return (start > 0 || end < length ? value.slice(start, end) : value);
+}
+
+function validateUrl(url, isConnect) {
+  var state = (isConnect ? 5 : 0);
+  var ch;
+  for (var i = 0; i < url.length; ++i) {
```

Comment:
I believe v8 is smart enough these days to produce the same asm as code that caches lengths, so this shouldn't be an issue.